### PR TITLE
transformer encoder

### PR DESCRIPTION
**Pull request by DevSinghSachan**
Hi Graham, Matthias,
I have added code for the 1 layer transformer encoder from Attention is all you need paper. This includes multi-head attention and Position feed forward layer.
Currently, masking is not done.

Thanks,

### CODE REVIEW

**Comment by msperber** [comment]
Sounds nice! I have a minor comment regarding code structure: Should we maybe move linear.Linear, mlp.MLP, transformer.AffineTransform, transformer.PositionwiseFeedForward to a general purpose "nn.py" module?
Also, the default dropout for the TransformerEncoder should probably be 0.0, to be consistent with the other components.

I haven't checked the network code in detail but hope to play around with it myself some when I find some time (not this week).

**Comment by DevSinghSachan** [comment]
Hi Matthias,
Thanks for the comments !
I have refactored the code into nn.py module and set the initiated the dropout value to 0.

I was thinking that in order to implement decoder module, we can do it with some changes to the attention Class. And that we can use directly compute the losses in the decode step for the entire sentence length using appropriate target masks. (suggested by @neubig )

So, I was thinking that should we write a new translation Class , something like TransformerTranslator() etc. and make changes the calc_loss() function.

**Comment by neubig** [comment]
Thanks @DevSinghSachan ! I'll try to take a look at this.

One thing I'd like you to try to be careful of is making a *minimum* of specialized functions. Any functions that are generalized and reusable, try to make them so. The best thing possible would be the ability to create `transformer.yaml` that simply used other reusable components.

That being said, I know this is not always possible, or not possible to do easily. In this case, if you need something like `TransformerTranslator`, let's add it to something like `specialized_translators.py` (similar to `specialized_encoder.py`). We can always modularize it later.

**Comment by philip30** [comment]
Hey @DevSinghSachan  this is minor, but can you make sure that you are using 2 tab space? I think it is not good to mix them up :)

File: xnmt/encoder.py [hunk 1]
```diff
@@ -311,6 +321,24 @@ def transduce(self, embed_sent):
   def initial_state(self):
     return PseudoState(self)
 
+
+class TransformerEncoder(BuilderEncoder, Serializable):
+  yaml_tag = u'!TransformerEncoder'
+
+  def __init__(self, context, input_dim=512, layers=1, hidden_dim=512, dropout=0.5, **kwargs):
```
Comment (msperber) [inline xnmt/encoder.py]: Sorry, actually what I meant regarding dropout would be something like:

def __init__(self, context, input_dim=512, layers=1, hidden_dim=512, dropout=None, **kwargs):
  self.dropout = dropout or context.dropout

That way we will be using the globally configured dropout, unless the TransformerEncoder-specific dropout value is set in the config.

File: examples/transformer.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+defaults:
```
Comment (msperber) [inline examples/transformer.yaml]: One more suggestion from my side would be to extend test/config/encoders.yaml to include a model with the new encoder (similar to this config but maybe with a tiny hidden size). Just to make sure future configs won't break this encoder.

Comment (DevSinghSachan) [reply examples/transformer.yaml]: That's a good idea. I will do this. thanks for the suggestions !!

File: test/test_beam_search.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from xnmt.translator import DefaultTranslator
 from xnmt.embedder import SimpleWordEmbedder
 from xnmt.encoder import LSTMEncoder
-from xnmt.attender import StandardAttender
```
Comment (neubig) [inline test/test_beam_search.py]: Fix to revert to StandarAttender

**Comment by DevSinghSachan** [comment]
Hi Graham and Matthias,

Transformer Encoder and Decoder along with masks support is completed.

Here are list of things to be done-
a)  Fix for label smoothing (code in separate branch)
b)  Modify `calc_loss` function in translator / decoder to use the output from TransformerDecoder.
c)  Write `generate` function for TransformerTranslator class.
d)  Write a class for special training procedure for Adam Optimizer as mentioned in 'AIAYN' paper.

**Comment by neubig** [comment]
Thanks! Have you tested to see if this gets reasonable accuracy?

**Comment by DevSinghSachan** [comment]
Hi Graham,

I have only done initial testing for the word train_loss/word. After 6 epochs using standard configurations, I am able to get a train_loss/word of 0.962 .
Currently, I working on getting BLEU score metric, in order to test the code.

**Comment by neubig** [comment]
OK, cool! That will be an important indicator of whether the code is correct or not, so I'll wait until we have an accuracy number before merging.

**Comment by msperber** [comment]
BTW, parts of the code could probably be simplified using the refactored masking code #189 , in case you want to make use of that.

**Comment by DevSinghSachan** [comment]
Hi Matt,
Thanks a lot for the suggestion!
I will look into the integration with refactored mask code once I initially get the Transformer model working.

**Comment by DevSinghSachan** [comment]
Currently, the performance of the model is not at pr with the standard example.
I am attaching the logs here.

Here is the log of the using a 1 layer model on the standard data set of ja->en
[TF.log](https://github.com/neulab/xnmt/files/1352246/TF.log)

Here is a sample of generated sentences.
[sent_gen.txt](https://github.com/neulab/xnmt/files/1352252/sent_gen.txt)

File: xnmt/expression_sequence.py [hunk 4]
```diff
@@ -32,7 +34,7 @@ def __len__(self):
     :returns: length of sequence
     """
     if self.expr_list: return len(self.expr_list)
-    else: return self.expr_tensor.dim()[0][0]
+    else: return self.expr_tensor.dim()[0][1]  # Earlier it was [0][0], but 0 is the number of dims
```
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/expression_sequence.py]: Sorry nevermind, you are correct.

**Pull request closed without merging**